### ISSUE TITLE
NMS Game Pass version not finding save data when correct file is selected

### ISSUE THREAD

**Issue by XDaast** [jaszhix/NoMansConnect#165]
I'm unable to load my save games from Xbox Game Pass.

**Comment by jaszhix** [comment]
Please fill out the issue or template - no way to fix otherwise because I don't have Game Pass.

**Comment by XDaast** [comment]
- Description of the problem
NoMansConnect stuck on loading screen and prompts that it cannot find save games.
- Windows version
Windows 10
- NMC version
PC Xbox Game Pass I can't find a patch version but it updated this morning.
- Absolute path to your storage_x.hg files.
I cannot locate that file. The save game path is supposed to be C:\Users\[user]\AppData\Local\Packages\HelloGames.NoMansSky_bs190hzg1sesy\SystemAppData\wgs\000901F1C2F741E7_29070100B936489ABCE8B9AF3980429C but this number is random for each player I believe. When I put it into NoMansConnect it still says it can't find game saves.
[NoMansConnectlogandjsonsettings.zip](https://github.com/jaszhix/NoMansConnect/files/4840605/NoMansConnectlogandjsonsettings.zip)


- Absolute path to your NMS install directory.
C:\Program Files\WindowsApps\HelloGames.NoMansSky_2.55.61121.0_x64__bs190hzg1sesy

Use Pastebin of directly attach these files (Users/<Username>/AppData/Roaming/NoMansConnect):

- NMC.log
- settings.json

- How can this bug be reproduced?
I suspect it has something to do with the way Xbox for PC works with the game.

Thanks for responding!

**Comment by lightblueandy** [comment]
Also seeing this issue. 

Saves are allegedly located here: **C:\Users\[user]\AppData\Local\Packages\HelloGames.NoMansSky_bs190hzg1sesy\SystemAppData\wgs\[randomIDfolder]\[randomIDfolder]\[randomFile]**  per issues and posts about other Windows Store games. 

I've tried copying the files to a different folder, point your app to them, even renaming (using the naming conventions in your code... save.hg, savenn.hg, storage, storagenn, etc) but the app doesn't recognize the files. 

I've attached the wgs folder from my NMS Windows store installation. Hopefully it helps.

[wgs.zip](https://github.com/jaszhix/NoMansConnect/files/4860262/wgs.zip)

OP of this issue also [https://github.com/goatfungus/NMSSaveEditor/issues/282](created an issue) in the referenced NMSSaveEditor project that included a save file (that OP did not include here). 

Thanks for your work on this and good luck!

**Comment by UltraCyberGod** [comment]
Minor but has anyone right click NMC and run as administrator just in case it's a permissions problem with the folder. I also expect that special permissions are set on those folders also. As it's basically a store product. Just trying to help...

**Comment by snoozbuster** [comment]
@jaszhix - I reverse-engineered most of the XGP save format and posted a writeup in [this issue](https://github.com/goatfungus/NMSSaveEditor/issues/306). not sure if you're still actively maintaining this tool but if so you should be able to use it to add support for XGP.

**Comment by jaszhix** [comment]
@snoozbuster Still maintaining it. A couple people in the related issues seem to be indicating it's not easy to do. This might break Microsoft's TOS to reverse engineer an encrypted UWP format. Might also need to change file ownership permissions to be able to read the UWP files. UWP is tightly integrated into Windows. It's one thing to reverse engineer a non-encrypted save format, but this might entail modifying system files.

I think for now I'd prefer to only support Steam/GOG since most players use those clients. If someone figures out a way to extract the save files from those binaries without being at odds with MS, I don't mind looking at a way to incorporate it in NMC.

**Comment by snoozbuster** [comment]
@jaszhix no encryption or system files - I was able to take the saves posted by @lightblueandy and load them unmodified into my game just by renaming a couple things. The game binary is stored in `%ProgramFiles%\WindowsApps` and accessing that requires changing folder ownership, but the saves themselves are stored in `%LocalAppData%` and don't require any mucking around with that. 

There is some sort of compression or perhaps encryption applied to the save file itself - but it is applied by the game, not MS, and it's not user-specific. Some reports suggest it's the same compression used for the PS4 version of the game.

**Comment by jaszhix** [comment]
Regardless, the end result is the same - we have to modify system files (UWP user data) to read the binaries. Hello Games has taken steps to obfuscate the JSON save data since this project has started, which has mostly deterred me from keeping up with newer features in the game. I'm not really keen on playing cat and mouse with them, but I'm open to PRs.

**Comment by dudeisbrendan03** [comment]
> Regardless, the end result is the same - we have to modify system files (UWP user data) to read the binaries. Hello Games has taken steps to obfuscate the JSON save data since this project has started, which has mostly deterred me from keeping up with newer features in the game. I'm not really keen on playing cat and mouse with them, but I'm open to PRs.

I'm pretty sure this wasn't the choice of Hello Games but to make the save compatible with Xbox cloud saving.

**Comment by jaszhix** [comment]
It was their choice, it's their game, but didn't mean to say it was intentionally subversive. It just has resulted in the game being harder to support with this third party tool.

**Comment by GlassHouse5** [comment]
I'm new to this, but the issue is mainly "XGP does not use the correct filetype (.dds instead of .hg) or doesn't support the save editor in general," right? I've been at it for a while, but I have no idea what I'm doing. I have no experience in any sort of computer science, so if you respond, try to dumb it down a little bit. Thanks.

**Comment by LucasBonafe** [comment]
@jaszhix, still don't have support for XGP? [NMSSaveEditor](https://github.com/goatfungus/NMSSaveEditor) normally supports save files...

**Comment by jaszhix** [comment]
Happy to accept patches for this.

**Comment by LucasBonafe** [comment]
> Happy to accept patches for this.

How can I help? Wanna make a call to speak about?

**Comment by jaszhix** [comment]
@LucasBonafe [NMSSaveEditor](https://github.com/goatfungus/NMSSaveEditor) is closed sourced, but perhaps someone familiar with Java can decompile it, or someone could ask the author how he accomplished it. NMC has Frontiers compatibility thanks to someone dropping a gist snippet, which I adapted into a bug fix release. I'm not specially experienced with reverse engineering proprietary formats, but happy to try if 1) someone can point me in the right direction and 2) it does not mess with Windows system permissions as a solution (main reason I closed this thread).
